usr/src/uts/common/fs/zfs/zfs_dir.c

```diff
@@ -774,25 +789,49 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
 	return (0);
 }
 
+/*
+ * The match type in the code for this function should conform to:
+ *
+ * ------------------------------------------------------------------------
+ * fs type  | z_norm      | lookup type | match type
+ * ---------|-------------|-------------|----------------------------------
+ * CS !norm | 0           |           0 | exact
+ * CS  norm | formD       |           0 | MT_IGNORE_FORM
+ * CI !norm | upper       |   !ZCIEXACT | MT_IGNORE_FORM
+ * CI !norm | upper       |    ZCIEXACT | MT_IGNORE_FORM | MT_PRESERVE_CASE
+ * CI  norm | upper|formX |   !ZCIEXACT | MT_IGNORE_FORM
+ * CI  norm | upper|formX |    ZCIEXACT | MT_IGNORE_FORM | MT_PRESERVE_CASE
+ * CM !norm | upper       |    !ZCILOOK | MT_IGNORE_FORM | MT_PRESERVE_CASE
+ * CM !norm | upper       |     ZCILOOK | MT_IGNORE_FORM
+ * CM  norm | upper|formX |    !ZCILOOK | MT_IGNORE_FORM | MT_PRESERVE_CASE
+ * CM  norm | upper|formX |     ZCILOOK | MT_IGNORE_FORM
```

Comment:
Interface-wise, I think it would be easier to understand the interface I misunderstood you to be implementing originally: two flags each of which specifies one orthogonal aspect to be normalized/ignored:
`MT_NORMALIZE_UNICODE_FORM` and `MT_CASE_INSENSITIVE` (or perhaps named `MT_IGNORE_FORM` and `MT_IGNORE_CASE`)

The table would look the same as this except the `match type` would substitute: `MT_IGNORE_FORM | MT_PRESERVE_CASE` -> `MT_NORMALIZE_UNICODE_FORM`
`MT_IGNORE_FORM` -> `MT_NORMALIZE_UNICODE_FORM | MT_CASE_INSENSITIVE`

I think what you've done here is still a good improvement on the existing logic so I won't insist on this additional change, but I think it would further improve understandability of the code.